File: production_engine/experimental/hierarchical_rag/hierarchical_chunker.py
"""
Hierarchical Chunker: Transforms canonical educational inputs into dual-level
Parent-Child Evidence Stores.
"""

import re
from typing import List, Dict, Any, Optional

from pipeline_experiment.shared.schemas.canonical_models import (
    CanonicalEducationalInput, SemanticChunk, TimeSpan
)
from production_engine.experimental.hierarchical_rag.schemas import (
    ChildChunk, ParentChunk, HierarchicalEvidenceStore
)


class HierarchicalChunker:
    """Builds dual-level Parent-Child chunk hierarchy from transcripts and supporting text."""

    CHILD_WINDOW_WORDS = 75    # Micro-unit for high-precision citation (~30-45s)
    PARENT_WINDOW_WORDS = 400  # Macro-unit for narrative reasoning context (~3-5 mins)

    @classmethod
    def tokenize_keywords(cls, text: str) -> List[str]:
        words = re.findall(r"\b[a-zA-Z_]{4,}\b", text.lower())
        stopwords = {
            "this", "that", "with", "from", "have", "were", "what", "when", "where",
            "which", "there", "their", "about", "would", "could", "should", "going",
            "because", "these", "those", "being", "other"
        }
        return list(set([w for w in words if w not in stopwords]))[:10]

    @classmethod
    def build_store(cls, canonical: CanonicalEducationalInput, transcript_data: Optional[Dict[str, Any]] = None) -> HierarchicalEvidenceStore:
        parents: List[ParentChunk] = []
        children: List[ChildChunk] = []
        parent_map: Dict[str, ParentChunk] = {}
        child_map: Dict[str, ChildChunk] = {}

        # 1. Process Spoken Audio if transcript segments are available
        if transcript_data and "segments" in transcript_data:
            segments = transcript_data["segments"]
            
            # Step A: Build Child Chunks
            curr_words: List[str] = []
            curr_start = 0.0
            curr_end = 0.0
            child_idx = 1
            temp_children: List[Dict[str, Any]] = []

            for seg in segments:
                text = seg.get("text", "").strip()
                s_time = float(seg.get("start", 0.0))
                e_time = float(seg.get("end", 0.0))

                if not curr_words:
                    curr_start = s_time
                curr_end = e_time
                curr_words.append(text)

                if len(" ".join(curr_words).split()) >= cls.CHILD_WINDOW_WORDS:
                    c_text = " ".join(curr_words)
                    cid = f"C_{child_idx:03d}"
                    eid = f"E_CHILD_{child_idx:03d}"
                    temp_children.append({
                        "child_id": cid,
                        "evidence_id": eid,
                        "start": curr_start,
                        "end": curr_end,
                        "text": c_text,
                        "word_count": len(c_text.split()),
                        "keywords": cls.tokenize_keywords(c_text)
                    })
                    child_idx += 1
                    curr_words = []

            if curr_words:
                c_text = " ".join(curr_words)
                cid = f"C_{child_idx:03d}"
                eid = f"E_CHILD_{child_idx:03d}"
                temp_children.append({
                    "child_id": cid,
                    "evidence_id": eid,
                    "start": curr_start,
                    "end": curr_end,
                    "text": c_text,
                    "word_count": len(c_text.split()),
                    "keywords": cls.tokenize_keywords(c_text)
                })

            # Step B: Group Children into Parent Windows (~4-5 children per parent)
            parent_idx = 1
            curr_p_children: List[Dict[str, Any]] = []
            curr_p_words = 0

            for tc in temp_children:
                curr_p_children.append(tc)
                curr_p_words += tc["word_count"]

                if curr_p_words >= cls.PARENT_WINDOW_WORDS:
                    p_id = f"P_{parent_idx:02d}"
                    p_eid = f"E_PARENT_{parent_idx:02d}"
                    p_start = curr_p_children[0]["start"]
                    p_end = curr_p_children[-1]["end"]
                    p_text = " ".join([c["text"] for c in curr_p_children])
                    c_ids = [c["evidence_id"] for c in curr_p_children]

                    parent_obj = ParentChunk(
                        parent_id=p_id,
                        evidence_id=p_eid,
                        title=f"Lecture Segment {parent_idx} ({int(p_start//60)}m{int(p_start%60):02d}s - {int(p_end//60)}m{int(p_end%60):02d}s)",
                        source_type="TRANSCRIPT",
                        time_span=TimeSpan(start=p_start, end=p_end),
                        full_text=p_text,
                        word_count=len(p_text.split()),
                        child_ids=c_ids
                    )
                    parents.append(parent_obj)
                    parent_map[p_eid] = parent_obj

                    for c in curr_p_children:
                        c_obj = ChildChunk(
                            child_id=c["child_id"],
                            evidence_id=c["evidence_id"],
                            parent_id=p_eid,
                            source_type="TRANSCRIPT",
                            time_span=TimeSpan(start=c["start"], end=c["end"]),
                            text=c["text"],
                            word_count=c["word_count"],
                            keywords=c["keywords"]
                        )
                        children.append(c_obj)
                        child_map[c["evidence_id"]] = c_obj

                    parent_idx += 1
                    curr_p_children = []
                    curr_p_words = 0

            if curr_p_children:
                p_id = f"P_{parent_idx:02d}"
                p_eid = f"E_PARENT_{parent_idx:02d}"
                p_start = curr_p_children[0]["start"]
                p_end = curr_p_children[-1]["end"]
                p_text = " ".join([c["text"] for c in curr_p_children])
                c_ids = [c["evidence_id"] for c in curr_p_children]

                parent_obj = ParentChunk(
                    parent_id=p_id,
                    evidence_id=p_eid,
                    title=f"Lecture Segment {parent_idx} ({int(p_start//60)}m{int(p_start%60):02d}s - {int(p_end//60)}m{int(p_end%60):02d}s)",
                    source_type="TRANSCRIPT",
                    time_span=TimeSpan(start=p_start, end=p_end),
                    full_text=p_text,
                    word_count=len(p_text.split()),
                    child_ids=c_ids
                )
                parents.append(parent_obj)
                parent_map[p_eid] = parent_obj

                for c in curr_p_children:
                    c_obj = ChildChunk(
                        child_id=c["child_id"],
                        evidence_id=c["evidence_id"],
                        parent_id=p_eid,
                        source_type="TRANSCRIPT",
                        time_span=TimeSpan(start=c["start"], end=c["end"]),
                        text=c["text"],
                        word_count=c["word_count"],
                        keywords=c["keywords"]
                    )
                    children.append(c_obj)
                    child_map[c["evidence_id"]] = c_obj

        # 2. Process Slide / PDF Sections
        slide_text = canonical.supporting_materials_text or (canonical.raw_content if canonical.input_type == "PDF" else "")
        if slide_text and "--- Slide " in slide_text:
            slide_sections = [s.strip() for s in slide_text.split("--- Slide ") if s.strip()]
            for s_idx, sec in enumerate(slide_sections):
                lines = sec.split("\n")
                slide_num = lines[0].split("---")[0].strip() if "---" in lines[0] else str(s_idx + 1)
                sec_text = "\n".join(lines[1:] if "---" in lines[0] else lines).strip()
                
                p_id = f"P_SLIDE_{slide_num}"
                p_eid = f"E_PARENT_SLIDE_{slide_num}"
                c_id = f"C_SLIDE_{slide_num}"
                c_eid = f"E_SLIDE_{slide_num}"

                parent_obj = ParentChunk(
                    parent_id=p_id,
                    evidence_id=p_eid,
                    title=f"Slide {slide_num} Full Context",
                    source_type="SLIDE",
                    full_text=sec_text,
                    word_count=len(sec_text.split()),
                    child_ids=[c_eid]
                )
                child_obj = ChildChunk(
                    child_id=c_id,
                    evidence_id=c_eid,
                    parent_id=p_eid,
                    source_type="SLIDE",
                    text=sec_text[:400],
                    word_count=len(sec_text[:400].split()),
                    keywords=cls.tokenize_keywords(sec_text)
                )

                parents.append(parent_obj)
                children.append(child_obj)
                parent_map[p_eid] = parent_obj
                child_map[c_eid] = child_obj

        return HierarchicalEvidenceStore(
            input_id=canonical.input_id,
            title=canonical.title,
            parents=parents,
            children=children,
            parent_map=parent_map,
            child_map=child_map
        )

File: server/engine/evidence/hierarchicalChunker.js
/**
 * server/engine/evidence/hierarchicalChunker.js
 *
 * Faithful port of Speech_To_Text/production_engine/experimental/hierarchical_rag/hierarchical_chunker.py.
 * Builds dual-level Parent-Child chunk hierarchy from transcripts, slides, and code snippets.
 * - Child Window: 75 words (micro-unit for high-precision citation, ~30-45s)
 * - Parent Window: 400 words (macro-unit for narrative reasoning context, ~3-5 mins)
 */

'use strict';

const CHILD_WINDOW_WORDS = 75;
const PARENT_WINDOW_WORDS = 400;

const STOPWORDS = new Set([
  'this', 'that', 'with', 'from', 'have', 'were', 'what', 'when', 'where',
  'which', 'there', 'their', 'about', 'would', 'could', 'should', 'going',
  'because', 'these', 'those', 'being', 'other'
]);

function tokenizeKeywords(text = '') {
  if (!text) return [];
  const words = text.toLowerCase().match(/\b[a-zA-Z_]{4,}\b/g) || [];
  const unique = Array.from(new Set(words.filter(w => !STOPWORDS.has(w))));
  return unique.slice(0, 10);
}

function countWords(text = '') {
  if (!text) return 0;
  return text.trim().split(/\s+/).filter(Boolean).length;
}

class HierarchicalChunker {
  /**
   * Build dual-level Parent-Child evidence store from session inputs.
   * @param {Object} sessionInputs - { voiceTranscript, documentTexts, codeSnippets, transcriptSegments, sessionId, title }
   * @returns {Object} HierarchicalEvidenceStore
   */
  static buildStore(sessionInputs = {}) {
    const parents = [];
    const children = [];
    const parentMap = {};
    const childMap = {};

    const inputId = sessionInputs.sessionId || 'session_' + Date.now();
    const title = sessionInputs.title || sessionInputs.subject || 'Educational Session';

    // ──────────────────────────────────────────────────────────────────────────
    // 1. Process Spoken Audio Transcript
    // ──────────────────────────────────────────────────────────────────────────
    const voiceText = (sessionInputs.voiceTranscript || '').trim();
    const rawSegments = sessionInputs.transcriptSegments || [];

    let segmentsToProcess = [];

    if (Array.isArray(rawSegments) && rawSegments.length > 0) {
      segmentsToProcess = rawSegments;
    } else if (voiceText.length > 0) {
      // Split raw transcript into natural pseudo-sentence segments to preserve incremental chunking
      const sentences = voiceText.match(/[^.!?]+[.!?]+|[^.!?]+$/g) || [voiceText];
      let runningTime = 0.0;
      for (const sent of sentences) {
        const text = sent.trim();
        if (!text) continue;
        const wCount = countWords(text);
        // Estimate ~130 words per minute -> ~0.46s per word
        const duration = Math.max(1.0, wCount * 0.46);
        segmentsToProcess.push({
          text,
          start: runningTime,
          end: runningTime + duration
        });
        runningTime += duration;
      }
    }

    if (segmentsToProcess.length > 0) {
      // Step A: Build Child Chunks (75 words each)
      let currWords = [];
      let currStart = 0.0;
      let currEnd = 0.0;
      let childIdx = 1;
      const tempChildren = [];

      for (const seg of segmentsToProcess) {
        const text = (seg.text || '').trim();
        const sTime = typeof seg.start === 'number' ? seg.start : 0.0;
        const eTime = typeof seg.end === 'number' ? seg.end : 0.0;

        if (currWords.length === 0) {
          currStart = sTime;
        }
        currEnd = eTime;
        currWords.push(text);

        if (countWords(currWords.join(' ')) >= CHILD_WINDOW_WORDS) {
          const cText = currWords.join(' ');
          const cid = `C_${String(childIdx).padStart(3, '0')}`;
          const eid = `E_CHILD_${String(childIdx).padStart(3, '0')}`;
          tempChildren.push({
            childId: cid,
            evidenceId: eid,
            start: currStart,
            end: currEnd,
            text: cText,
            wordCount: countWords(cText),
            keywords: tokenizeKeywords(cText)
          });
          childIdx++;
          currWords = [];
        }
      }

      if (currWords.length > 0) {
        const cText = currWords.join(' ');
        const cid = `C_${String(childIdx).padStart(3, '0')}`;
        const eid = `E_CHILD_${String(childIdx).padStart(3, '0')}`;
        tempChildren.push({
          childId: cid,
          evidenceId: eid,
          start: currStart,
          end: currEnd,
          text: cText,
          wordCount: countWords(cText),
          keywords: tokenizeKeywords(cText)
        });
      }

      // Step B: Group Children into Parent Windows (~400 words, ~4-5 children)
      let parentIdx = 1;
      let currPChildren = [];
      let currPWords = 0;

      for (const tc of tempChildren) {
        currPChildren.push(tc);
        currPWords += tc.wordCount;

        if (currPWords >= PARENT_WINDOW_WORDS) {
          const pId = `P_${String(parentIdx).padStart(2, '0')}`;
          const pEid = `E_PARENT_${String(parentIdx).padStart(2, '0')}`;
          const pStart = currPChildren[0].start;
          const pEnd = currPChildren[currPChildren.length - 1].end;
          const pText = currPChildren.map(c => c.text).join(' ');
          const cIds = currPChildren.map(c => c.evidenceId);

          const startMin = Math.floor(pStart / 60);
          const startSec = Math.floor(pStart % 60);
          const endMin = Math.floor(pEnd / 60);
          const endSec = Math.floor(pEnd % 60);

          const parentObj = {
            parentId: pId,
            evidenceId: pEid,
            title: `Lecture Segment ${parentIdx} (${startMin}m${String(startSec).padStart(2, '0')}s - ${endMin}m${String(endSec).padStart(2, '0')}s)`,
            sourceType: 'TRANSCRIPT',
            timeSpan: { start: pStart, end: pEnd },
            fullText: pText,
            wordCount: countWords(pText),
            childIds: cIds
          };
          parents.push(parentObj);
          parentMap[pEid] = parentObj;

          for (const c of currPChildren) {
            const childObj = {
              childId: c.childId,
              evidenceId: c.evidenceId,
              parentId: pEid,
              sourceType: 'TRANSCRIPT',
              timeSpan: { start: c.start, end: c.end },
              text: c.text,
              wordCount: c.wordCount,
              keywords: c.keywords
            };
            children.push(childObj);
            childMap[c.evidenceId] = childObj;
          }

          parentIdx++;
          currPChildren = [];
          currPWords = 0;
        }
      }

      if (currPChildren.length > 0) {
        const pId = `P_${String(parentIdx).padStart(2, '0')}`;
        const pEid = `E_PARENT_${String(parentIdx).padStart(2, '0')}`;
        const pStart = currPChildren[0].start;
        const pEnd = currPChildren[currPChildren.length - 1].end;
        const pText = currPChildren.map(c => c.text).join(' ');
        const cIds = currPChildren.map(c => c.evidenceId);

        const startMin = Math.floor(pStart / 60);
        const startSec = Math.floor(pStart % 60);
        const endMin = Math.floor(pEnd / 60);
        const endSec = Math.floor(pEnd % 60);

        const parentObj = {
          parentId: pId,
          evidenceId: pEid,
          title: `Lecture Segment ${parentIdx} (${startMin}m${String(startSec).padStart(2, '0')}s - ${endMin}m${String(endSec).padStart(2, '0')}s)`,
          sourceType: 'TRANSCRIPT',
          timeSpan: { start: pStart, end: pEnd },
          fullText: pText,
          wordCount: countWords(pText),
          childIds: cIds
        };
        parents.push(parentObj);
        parentMap[pEid] = parentObj;

        for (const c of currPChildren) {
          const childObj = {
            childId: c.childId,
            evidenceId: c.evidenceId,
            parentId: pEid,
            sourceType: 'TRANSCRIPT',
            timeSpan: { start: c.start, end: c.end },
            text: c.text,
            wordCount: c.wordCount,
            keywords: c.keywords
          };
          children.push(childObj);
          childMap[c.evidenceId] = childObj;
        }
      }
    }

    // ──────────────────────────────────────────────────────────────────────────
    // 2. Process Slide / PDF Sections & Supporting Materials
    // ──────────────────────────────────────────────────────────────────────────
    const docArray = Array.isArray(sessionInputs.documentTexts) ? sessionInputs.documentTexts : [];
    const docText = docArray.join('\n\n');

    if (docText.includes('--- Slide ') || docArray.length > 0) {
      let slideSections = [];
      if (docText.includes('--- Slide ')) {
        slideSections = docText.split('--- Slide ').map(s => s.trim()).filter(Boolean);
      } else {
        // Each document element or markdown header section treated as slide unit
        slideSections = docArray.map(d => d.trim()).filter(Boolean);
      }

      for (let sIdx = 0; sIdx < slideSections.length; sIdx++) {
        const sec = slideSections[sIdx];
        const lines = sec.split('\n');
        let slideNum = String(sIdx + 1);
        let secText = sec;

        if (lines[0].includes('---')) {
          slideNum = lines[0].split('---')[0].trim() || slideNum;
          secText = lines.slice(1).join('\n').trim();
        }

        const secWords = countWords(secText);
        if (secWords > PARENT_WINDOW_WORDS) {
          // Chunk long document/PDF into standard 400-word parent windows and 75-word child chunks
          const words = secText.split(/\s+/).filter(Boolean);
          let pIdx = 1;
          for (let wPos = 0; wPos < words.length; wPos += PARENT_WINDOW_WORDS) {
            const pWords = words.slice(wPos, wPos + PARENT_WINDOW_WORDS);
            const pId = `P_DOC_${slideNum}_${pIdx}`;
            const pEid = `E_PARENT_DOC_${slideNum}_${pIdx}`;
            const pText = pWords.join(' ');

            const pChildIds = [];
            let cIdx = 1;
            for (let cPos = 0; cPos < pWords.length; cPos += CHILD_WINDOW_WORDS) {
              const cWords = pWords.slice(cPos, cPos + CHILD_WINDOW_WORDS);
              const cText = cWords.join(' ');
              const cId = `C_DOC_${slideNum}_${pIdx}_${cIdx}`;
              const cEid = `E_DOC_${slideNum}_${pIdx}_${cIdx}`;
              pChildIds.push(cEid);

              const childObj = {
                childId: cId,
                evidenceId: cEid,
                parentId: pEid,
                sourceType: 'SLIDE',
                text: cText,
                wordCount: countWords(cText),
                keywords: tokenizeKeywords(cText)
              };
              children.push(childObj);
              childMap[cEid] = childObj;
              cIdx++;
            }

            const parentObj = {
              parentId: pId,
              evidenceId: pEid,
              title: `Document Section ${slideNum}.${pIdx} Context`,
              sourceType: 'SLIDE',
              fullText: pText,
              wordCount: countWords(pText),
              childIds: pChildIds
            };
            parents.push(parentObj);
            parentMap[pEid] = parentObj;
            pIdx++;
          }
        } else {
          const pId = `P_SLIDE_${slideNum}`;
          const pEid = `E_PARENT_SLIDE_${slideNum}`;
          const cId = `C_SLIDE_${slideNum}`;
          const cEid = `E_SLIDE_${slideNum}`;

          const parentObj = {
            parentId: pId,
            evidenceId: pEid,
            title: `Slide ${slideNum} Full Context`,
            sourceType: 'SLIDE',
            fullText: secText,
            wordCount: countWords(secText),
            childIds: [cEid]
          };

          const childSnippet = secText.substring(0, 400);
          const childObj = {
            childId: cId,
            evidenceId: cEid,
            parentId: pEid,
            sourceType: 'SLIDE',
            text: childSnippet,
            wordCount: countWords(childSnippet),
            keywords: tokenizeKeywords(secText)
          };

          parents.push(parentObj);
          children.push(childObj);
          parentMap[pEid] = parentObj;
          childMap[cEid] = childObj;
        }
      }
    }

    // ──────────────────────────────────────────────────────────────────────────
    // 3. Process Code Snippets
    // ──────────────────────────────────────────────────────────────────────────
    const codeText = (sessionInputs.codeSnippets || '').trim();
    if (codeText.length > 0) {
      const pId = 'P_CODE_01';
      const pEid = 'E_PARENT_CODE_01';
      const cId = 'C_CODE_01';
      const cEid = 'E_CODE_01';

      const parentObj = {
        parentId: pId,
        evidenceId: pEid,
        title: 'Code Snippets Context',
        sourceType: 'CODE',
        fullText: codeText,
        wordCount: countWords(codeText),
        childIds: [cEid]
      };

      const childObj = {
        childId: cId,
        evidenceId: cEid,
        parentId: pEid,
        sourceType: 'CODE',
        text: codeText.substring(0, 400),
        wordCount: countWords(codeText.substring(0, 400)),
        keywords: tokenizeKeywords(codeText)
      };

      parents.push(parentObj);
      children.push(childObj);
      parentMap[pEid] = parentObj;
      childMap[cEid] = childObj;
    }

    return {
      inputId,
      title,
      parents,
      children,
      parentMap,
      childMap
    };
  }
}

module.exports = {
  HierarchicalChunker,
  CHILD_WINDOW_WORDS,
  PARENT_WINDOW_WORDS,
  tokenizeKeywords
};
